### PR TITLE
Provide Mojo services for passing application-specific messages betwe…

### PR DESCRIPTION
**Pull request by jason-simmons**
…en the host and Flutter

### CODE REVIEW

File: sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java [hunk 1]
```diff
@@ -479,4 +495,63 @@ private boolean handleAccessibilityHoverEvent(MotionEvent event) {
...
+    public void setOnMessageListener(OnMessageListener listener) {
+        mOnMessageListener = listener;
+    }
+
+    public interface OnMessageListener {
+        /**
+         * Called when a message is received.
+         * @return the reply to the message (can be null)
+         */
+        String onMessage(String message);
```
Comment (abarth) [inline sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java]: It's slightly limiting for this to require the reply synchronously but I can see how that's much less verbose in Java.

File: sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java [hunk 2]
```diff
@@ -326,6 +330,13 @@ public void connectToService(Context context, Core core, MessagePipeHandle pipe)
                 RawKeyboardService.MANAGER.bind(new RawKeyboardServiceImpl(mRawKeyboardState), pipe);
             }
         });
+
+        registry.register(HostApplication.MANAGER.getName(), new ServiceFactory() {
+            @Override
+            public void connectToService(Context context, Core core, MessagePipeHandle pipe) {
+                HostApplication.MANAGER.bind(new HostApplicationImpl(), pipe);
```
Comment (abarth) [inline sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java]: We could use this inner class trick for these other two local services too.

File: sky/services/platform/host_app.mojom [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+interface HostApplication {
+  SendToHost(string message) => (string? reply);
+};
+
+/// A service provided by the Flutter framework that can receive messages
+/// sent by the host.
+[ServiceName="flutter::platform::EmbeddedApplication"]
+interface EmbeddedApplication {
+  SendToFlutter(string message) => (string? reply);
+};
```
Comment (abarth) [inline sky/services/platform/host_app.mojom]: Is there a reason to have two different interfaces?  Why not just have one?  You could then call "SendToHost" and "SendToFlutter" just "SendString"

**Comment by abarth** [comment]
LGTM, but I'd probably merge the two interfaces into one.

**Comment by j9brown** [comment]
More common idiom in Android is to have a method which takes a Bundle and
returns a Bundle.  This allows passing structured data as a map, including
the potential of passing Binder objects.

Now it might not make sense to pass Binder objects around but it might
still be worth considering the use of a Map of some kind rather than a
String.  (Though Mojo does you no favors here since Mojo maps are
homogeneous.)

Anyhow, worth seeing how clients use this API and adapting accordingly.  It
might also be quite reasonable to simply transport Mojo message pipes.

File: sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java (hunk 1 above)
Comment (krisgiesing) [inline sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java]: Actually I think all that's required is to pass the SendToHostResponse down to the message listener. The API just above this one is already structured asynchronously.

**Comment by krisgiesing** [comment]
I'm not sure String vs. Map matters too much because people can just JSON encode.

A slightly different consideration, harder to deal with, is clients that want a streaming interface.  I guess in that case you could write a customized Mojo interface.

**Comment by j9brown** [comment]
Well, or clients that want to pass capabilities to bootstrap other
capabilities (ie. mojo handles).  That's what made me think of Bundle.

**Comment by jason-simmons** [comment]
Merged the interfaces into one ApplicationMessages interface, and added a message name argument.  Host apps can register synchronous or async handlers for a given message name.

PTAL

File: sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java [hunk 4]
```diff
@@ -479,4 +500,103 @@ private boolean handleAccessibilityHoverEvent(MotionEvent event) {
...
+            if (asyncListener != null) {
+                asyncListener.onMessage(message, new MessageResponseAdapter(callback));
+                return;
+            }
+
+            callback.call(null);
+        }
+    }
+
+    private static class MessageResponseAdapter implements MessageResponse {
```
Comment (krisgiesing) [inline sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java]: Can you elaborate on the need for the adapter in a comment?

Comment (jason-simmons) [inline sky/shell/platform/android/org/domokit/sky/shell/PlatformViewAndroid.java]: done

**Comment by krisgiesing** [comment]
LGTM

**Pull request merged**